Add release note for 2.8.0 (#10874)
apache/pulsar: site2/website/release-notes.md

- [PIP 83] Message consumption with pooled buffer [#10184](https://github.com/apache/pulsar/pull/10184)
- [PIP 85] Support get reader schema for a message in Java Client [#10476](https://github.com/apache/pulsar/pull/10476)

#### Transactions
- Support pending ack state persistence [#8881](https://github.com/apache/pulsar/pull/8881)
- Stable position and low watermark for the transaction buffer [#9195](https://github.com/apache/pulsar/pull/9195)
- Transaction timeout support [#9229](https://github.com/apache/pulsar/pull/9229)
- Transaction coordinator retry to complete the transaction [#9236](https://github.com/apache/pulsar/pull/9236)
- Fix race condition when appending transaction log [#9238](https://github.com/apache/pulsar/pull/9238)
- Transaction buffer snapshot [#9490](https://github.com/apache/pulsar/pull/9490)
- Add metrics for transaction coordinator [#9706](https://github.com/apache/pulsar/pull/9706)
- Clean the useless transaction individual acknowledgements based on low watermark [#9722](https://github.com/apache/pulsar/pull/9722)

> - Fix memory leak when delete transaction marker [#9751](https://github.com/apache/pulsar/pull/9751)

Reviewer: ```suggestion
- Fix memory leak when deleting transaction marker [#9751](https://github.com/apache/pulsar/pull/9751)
```

- Check the transaction state at the client side [#9776](https://github.com/apache/pulsar/pull/9776)
- Clean aborted transactions for the transaction buffer [#9974](https://github.com/apache/pulsar/pull/9974)
- Fix transaction coordinator retry to end transaction [#10131](https://github.com/apache/pulsar/pull/10131)

> - Fix NPE when open a new transaction [#10139](https://github.com/apache/pulsar/pull/10139)

Reviewer: ```suggestion
- Fix NPE when opening a new transaction [#10139](https://github.com/apache/pulsar/pull/10139)
```

- Fix transaction coordinator recover timeout [#10162](https://github.com/apache/pulsar/pull/10162)
- Handling committing and aborting state when recovering transaction coordinator [#10179](https://github.com/apache/pulsar/pull/10179)
- Fix NPE in case of enableTransactionCoordinator=false [#10182](https://github.com/apache/pulsar/pull/10182)
- Fix transaction buffer client timeout [#10206](https://github.com/apache/pulsar/pull/10206)
- Fix recover max local id issue for the transaction coordinator [#10224](https://github.com/apache/pulsar/pull/10224)
- Support using transactions on standalone [#10238](https://github.com/apache/pulsar/pull/10238)
- Fix transaction buffer lookup issue [#10257](https://github.com/apache/pulsar/pull/10257)
- Fix transaction timeout issue at the client side [#10323](https://github.com/apache/pulsar/pull/10323)
- Fix transaction client reconnect issue after transaction coordinator unloaded [#10327](https://github.com/apache/pulsar/pull/10327)
- Fix transaction timeout not canceled after expired [#10366](https://github.com/apache/pulsar/pull/10366)
- Transaction log low watermark optimization [#10422](https://github.com/apache/pulsar/pull/10422)

> - Fix transaction buffer client channel not active problem [#10407](https://github.com/apache/pulsar/pull/10407)

Reviewer: ```suggestion
- Fix the issue of transaction buffer client channel is inactive [#10407](https://github.com/apache/pulsar/pull/10407)
```

- Admin API for getting transaction coordinator stats [#10639](https://github.com/apache/pulsar/pull/10639)
- Admin API for getting transaction in transaction buffer stats [#10642](https://github.com/apache/pulsar/pull/10642)
- Admin API for getting transaction in pending ack stats [#10648](https://github.com/apache/pulsar/pull/10648)
- Admin API for getting transaction buffer stats and pending ack stats [#10650](https://github.com/apache/pulsar/pull/10650)

> - Admin API for getting transaction coordinator internal stats [#10653](https://github.com/apache/pulsar/pull/10653)

Reviewer: Add the word "Add" before "Admin API ...."

- Setup transaction metadata with metadata store [#10677](https://github.com/apache/pulsar/pull/10677)
- Fix issue with acknowledge messages from multiple subscriptions of a topic [#10689](https://github.com/apache/pulsar/pull/10689)
- Admin API for getting transaction metadata [#10690](https://github.com/apache/pulsar/pull/10690)
- Admin API for getting slow transactions [#10701](https://github.com/apache/pulsar/pull/10701)
- Fix transaction log handle managed ledger WriteFail state [#10711](https://github.com/apache/pulsar/pull/10711)
- Admin API for getting pending ack internal stats [#10725](https://github.com/apache/pulsar/pull/10725)
- Fix transaction ack delete marker position when don't have transaction ack [#10741](https://github.com/apache/pulsar/pull/10741)
- Fix transaction pending ack generate managedLedgerStats fail exception [#10797](https://github.com/apache/pulsar/pull/10797)

> - Use zookeeper prometheus metric provider to export zookeeper metrics [#10803](https://github.com/apache/pulsar/pull/10803)

Reviewer: ```suggestion
- Use Zookeeper Prometheus metric provider to export Zookeeper metrics [#10803](https://github.com/apache/pulsar/pull/10803)
```

#### Security

> - Optional auth method name header in http authentication [#6799](https://github.com/apache/pulsar/pull/6799)

Reviewer: ```suggestion
- Optional auth method name header in HTTP authentication [#6799](https://github.com/apache/pulsar/pull/6799)
```

- Topics level authorization support [#7523](https://github.com/apache/pulsar/pull/7523)
- Authorization for setting topic/subscription auto-creation and subscription expire time on namespace [#7900](https://github.com/apache/pulsar/pull/7900)
- Allow serializable stream-provider field into AuthenticationTls [#10020](https://github.com/apache/pulsar/pull/10020)
- Add configuration for running OWASP Dependency Check for all modules [#10288](https://github.com/apache/pulsar/pull/10288)

> - Update default TLS protocols to TLSv1.3,TLSv1.2 for the broker and proxy [#10598](https://github.com/apache/pulsar/pull/10598)

Reviewer: ```suggestion
- Update default TLS protocols to TLSv1.3 and TLSv1.2 for the broker and proxy [#10598](https://github.com/apache/pulsar/pull/10598)
```

> #### Broker

Reviewer 2: I think it is worth adding this PR to the broker's list of changes: https://github.com/apache/pulsar/pull/10223. It introduces a deprecated field as well as the preferred field. No fields will be removed until 3.0.0 (presumably), but I think we should advertise the change/deprecation here.

Author: @michaeljmarshall Added in both broker and client.

> - Share EventLoopGroup between Broker and BookKeeper client [#2603](https://github.com/apache/pulsar/pull/2603)

Reviewer: ```suggestion
- Share EventLoopGroup between broker and BookKeeper client [#2603](https://github.com/apache/pulsar/pull/2603)
```

- Dispatch batch messages according to consumer permits [7266](https://github.com/apache/pulsar/pull/7266)
- Improve the max pending publish buffer mechanism [7406](https://github.com/apache/pulsar/pull/7406)
- Allow disabling HTTP TRACE/TRACK verbs [#7907](https://github.com/apache/pulsar/pull/7907)
- Fix DispatchRateLimiter does not take effect [#8611](https://github.com/apache/pulsar/pull/8611)
- Validate namespace isolation policy regex before updating [#8804](https://github.com/apache/pulsar/pull/8804)
- Perform automatically cert refresh for pulsar-admin [#8831](https://github.com/apache/pulsar/pull/8831)

> - Fix Updating rack info dynamically [#8844](https://github.com/apache/pulsar/pull/8844)

Reviewer: ```suggestion
- Fix updating rack info dynamically [#8844](https://github.com/apache/pulsar/pull/8844)
```

- Fix deadlock when checking message expiration [#8877](https://github.com/apache/pulsar/pull/8877)
- Support limit max topics per namespace [#8942](https://github.com/apache/pulsar/pull/8942)
- Make ledger rollover check task internally [#8946](https://github.com/apache/pulsar/pull/8946)
- Clean up topic that failed to unload from the cache [#8968](https://github.com/apache/pulsar/pull/8968)
- Support get broker response for the message acknowledgement [#8996](https://github.com/apache/pulsar/pull/8996)
- Support message index for a topic [#9039](https://github.com/apache/pulsar/pull/9039)
- Supporting limit max topics per namespace by namespace policy [#9042](https://github.com/apache/pulsar/pull/9042)
- Streaming dipsatcher support [#9056](https://github.com/apache/pulsar/pull/9056)
- Make Netty acceptor threadPool size configurable [#9061](https://github.com/apache/pulsar/pull/9061)
- Fix deadlock when unloading namespace bundles [#9116](https://github.com/apache/pulsar/pull/9116)
- Fixed checking for maxTopicsPerNamespace [#9121](https://github.com/apache/pulsar/pull/9121)

> - Change the service URL to a not required param when create a cluster [#9127](https://github.com/apache/pulsar/pull/9127)

Reviewer: ```suggestion
- Change the service URL to a not required param when creating a cluster [#9127](https://github.com/apache/pulsar/pull/9127)
```

- Support setting replicator dispatch rate policy at the topic level [#9175](https://github.com/apache/pulsar/pull/9175)
- Fix max topic for namespace does not work [#9193](https://github.com/apache/pulsar/pull/9193)
- Fix memory leak of the managed ledger interceptor [#9194](https://github.com/apache/pulsar/pull/9194)
- Fix max consumers per topic cannot be disabled at the namespace level [#9214](https://github.com/apache/pulsar/pull/9214)

> - Support schema rest api for V1 topics [#9218](https://github.com/apache/pulsar/pull/9218)

Reviewer: ```suggestion
- Support schema REST API for V1 topics [#9218](https://github.com/apache/pulsar/pull/9218)
```

> - Fix peek message metadata while enable broker entry metadata [#9255](https://github.com/apache/pulsar/pull/9255)

Reviewer: ```suggestion
- Fix peek message metadata while enabling broker entry metadata [#9255](https://github.com/apache/pulsar/pull/9255)
```

- Support fetching metadata from entry data in publish callback [#9257](https://github.com/apache/pulsar/pull/9257)
- Wait for the async broker port listener close operations to complete at shutdown [#9308](https://github.com/apache/pulsar/pull/9308)
- Support script based mapping network topology [#9363](https://github.com/apache/pulsar/pull/9363)
- Make managed ledger storage configurable [#9397](https://github.com/apache/pulsar/pull/9397)
- Support setting enabled subscription types [#9401](https://github.com/apache/pulsar/pull/9401)
- Fixed NPE and cache invalidation in leader election [#9460](https://github.com/apache/pulsar/pull/9460)

> - Fix exception when get optional field for Protobuf message [#9468](https://github.com/apache/pulsar/pull/9468)

Reviewer: ```suggestion
- Fix exception when get an optional field for Protobuf message [#9468](https://github.com/apache/pulsar/pull/9468)
```

- Ignore replicated subscription configurations from the client when disabled by broker [#9523](https://github.com/apache/pulsar/pull/9523)

> - Support expose producer metrics through prometheus endpoint [#9541](https://github.com/apache/pulsar/pull/9541)

Reviewer: ```suggestion
- Support expose producer metrics through Prometheus endpoint [#9541](https://github.com/apache/pulsar/pull/9541)
```

- Fix NPE in unblock stuck subscription task when dispatcher has not created [#10430](https://github.com/apache/pulsar/pull/10430)
- Fix topic loading fails without any error when replicator init fails [#10432](https://github.com/apache/pulsar/pull/10432)
- Set timeout to unblock web-threads on update partition API [#10447](https://github.com/apache/pulsar/pull/10447)
- Fix CPU 100% when deleting namespace [#10454](https://github.com/apache/pulsar/pull/10454)
- Remove residual info after forcibly deleting the namespace [#10465](https://github.com/apache/pulsar/pull/10465)
- Fix potential memory leak of TopicPolicies [#10466](https://github.com/apache/pulsar/pull/10466)
- Fix publish callback's entry data is null during ledger rollover [#10467](https://github.com/apache/pulsar/pull/10467)
- Fix readModifyUpdate should return the new value in the future for metadata cache [#10474](https://github.com/apache/pulsar/pull/10474)
- Fix partitioned system topic check issue [#10529](https://github.com/apache/pulsar/pull/10529)
- Removed AspectJ based metrics for ZooKeeper [#10533](https://github.com/apache/pulsar/pull/10533)

> - Allow to create MetadataCache with custom serde [#10543](https://github.com/apache/pulsar/pull/10543)

Reviewer: ```suggestion
- Allow creatintg MetadataCache with custom serde [#10543](https://github.com/apache/pulsar/pull/10543)
```

- Fix ack receipt version check issue [#10551](https://github.com/apache/pulsar/pull/10551)
- Expose average message size metrics for a topic [#10553](https://github.com/apache/pulsar/pull/10553)
- Fixed missed ZK caching when fetching list of namespaces for a tenant [#10594](https://github.com/apache/pulsar/pull/10594)
- Setup pulsar cluster with MetadataStore [#10600](https://github.com/apache/pulsar/pull/10600)
- Setup initial namespaces with MetadataStore [#10612](https://github.com/apache/pulsar/pull/10612)
- Convert bundle split into an HTTP async operation [#10619](https://github.com/apache/pulsar/pull/10619)
- Add metrics for non-contiguous deleted messages range [#10638](https://github.com/apache/pulsar/pull/10638)
- Fix consumer related topic stats only available while consumer or reader are connected [#10644](https://github.com/apache/pulsar/pull/10644)

> - Allow to configure the number of BK client worker threads [#10649](https://github.com/apache/pulsar/pull/10649)

Reviewer: ```suggestion
- Allow configuring the number of BK client worker threads [#10649](https://github.com/apache/pulsar/pull/10649)
```

- Fix ConcurrentOpenLongPairRangeSet remove all ranges [#10656](https://github.com/apache/pulsar/pull/10656)
- Ensure all the ReadHandle gets properly closed on cache invalidation [#10659](https://github.com/apache/pulsar/pull/10659)
- Avoid the context switch when managedLedgerNewEntriesCheckDelayInMillis=0 [#10660](https://github.com/apache/pulsar/pull/10660)

> - Allow to configure busy-wait in broker and client [#10661](https://github.com/apache/pulsar/pull/10661)

Reviewer: ```suggestion
- Allow configuring busy-wait in broker and client [#10661](https://github.com/apache/pulsar/pull/10661)
```

> - Use single Jwt parser instance [#10664](https://github.com/apache/pulsar/pull/10664)

Reviewer: ```suggestion
- Use a single JWT parser instance [#10664](https://github.com/apache/pulsar/pull/10664)
```

- Fix issues in advanceNonDurableCursors [#10667](https://github.com/apache/pulsar/pull/10667)

> - LockManager use concurrent hash map to handle locks notifiications [#10680](https://github.com/apache/pulsar/pull/10680)

Reviewer: ```suggestion
- LockManager use a concurrent hash map to handle locks notifications [#10680](https://github.com/apache/pulsar/pull/10680)
```

- Avoid making copies of internal maps when iterating [#10691](https://github.com/apache/pulsar/pull/10691)
- Fix issue where StackOverflowError occurs when trying to redeliver a large number of already acked messages [#10696](https://github.com/apache/pulsar/pull/10696)
- Fix NPE when filtering read entries [#10704](https://github.com/apache/pulsar/pull/10704)

> - Fix deadlock when enable managed ledger interceptor [#10706](https://github.com/apache/pulsar/pull/10706)

Reviewer: ```suggestion
- Fix deadlock when enabling managed ledger interceptor [#10706](https://github.com/apache/pulsar/pull/10706)
```

#### Functions
- Update default function log location in function_worker.yaml [#8470](https://github.com/apache/pulsar/pull/8470)
- Optimize batch source discovery and task ack [#8498](https://github.com/apache/pulsar/pull/8498)
- Provide an interface for functions worker service [#8560](https://github.com/apache/pulsar/pull/8560)
- Improve naming of threads used in batch source [#8608](https://github.com/apache/pulsar/pull/8608)
- Fix the reader used to read the metadata topic during worker service initialization not been closed [#8637](https://github.com/apache/pulsar/pull/8637)
- Add timeout to hasMessageAvailable to leader election process [#8687](https://github.com/apache/pulsar/pull/8687)
- Support key based batch builder for Go Functions [#8761](https://github.com/apache/pulsar/pull/8761)
- Fix panic when discarding message for Go Functions [#8776](https://github.com/apache/pulsar/pull/8776)
- Move initialize dlog namespace metadata to bin/pulsar [#8781](https://github.com/apache/pulsar/pull/8781)
- Support URL fetching for Go/Python Functions [#8808](https://github.com/apache/pulsar/pull/8808)

> - Add cmd flag for retain key ordering [#8886](https://github.com/apache/pulsar/pull/8886)

Reviewer: ```suggestion
- Add cmd flag for retaining key ordering [#8886](https://github.com/apache/pulsar/pull/8886)
```

- Supports to use package command [#8973](https://github.com/apache/pulsar/pull/8973)
- Support create and update sink with package name [#8987](https://github.com/apache/pulsar/pull/8987)
- Support create and update source with package name [#8988](https://github.com/apache/pulsar/pull/8988)
- Allow stats operations not to be blocked in functions [#9005](https://github.com/apache/pulsar/pull/9005)
- Adding timeout to open table call for function state [#9006](https://github.com/apache/pulsar/pull/9006)
- Fix get function info error of REST API [#9115](https://github.com/apache/pulsar/pull/9115)

> - Improve the error message when create sinks [#9131](https://github.com/apache/pulsar/pull/9131)

Reviewer: ```suggestion
- Improve the error message when creating sinks [#9131](https://github.com/apache/pulsar/pull/9131)
```

- Expose Pulsar Admin through Function context [#9246](https://github.com/apache/pulsar/pull/9246)
- Enable Function Workers to use exclusive producer to write to internal topics [#9275](https://github.com/apache/pulsar/pull/9275)
- Add additional logging when setting up state table for function instance [#9304](https://github.com/apache/pulsar/pull/9304)
- Allow memory limit to be set for the pulsar client used in the ThreadRuntime in Pulsar Function [#9320](https://github.com/apache/pulsar/pull/9320)
- Make admin operations on Statestore non blocking [#9348](https://github.com/apache/pulsar/pull/9348)
- Fix maxPendingAsyncRequests not affect Kubernetes Runtime [#9349](https://github.com/apache/pulsar/pull/9349)
- Add download directory support to function Kubernetes Runtime [#9377](https://github.com/apache/pulsar/pull/9377)
- Optimize built-in source/sink startup by eliminating redundant NAR unpacking and checksum calculation [#9413](https://github.com/apache/pulsar/pull/9413) [#9500](https://github.com/apache/pulsar/pull/9500)

> - Enhance kubernetes manifest customizer with default options [#9445](https://github.com/apache/pulsar/pull/9445)

Reviewer: ```suggestion
- Enhance Kubernetes manifest customizer with default options [#9445](https://github.com/apache/pulsar/pull/9445)
```

- Fix possible deadlock on broker-function service startup [#9499](https://github.com/apache/pulsar/pull/9499)
- Close InputStreams properly [#9568](https://github.com/apache/pulsar/pull/9568)
- Add maximum allowed amount of resources setting for functions [#9584](https://github.com/apache/pulsar/pull/9584)
- Support writing general records to Pulsar sink [#9590](https://github.com/apache/pulsar/pull/9590)
- Move metrics port configuration to InstanceConfig [#9610](https://github.com/apache/pulsar/pull/9610)
- Add resource granularity settings for functions [#9736](https://github.com/apache/pulsar/pull/9736)
- Prevent invalid broker or proxy configuration for authorization [#9746](https://github.com/apache/pulsar/pull/9746)
- Log stacktraces of threads that failed to terminate on shutdown within timeout in ExecutorProvider [#9840](https://github.com/apache/pulsar/pull/9840)
- Support get partition index for a Record [#9947](https://github.com/apache/pulsar/pull/9947)
- Deleting a Pulsar Function with a name that includes a colon character crashes the pulsar broker [#9946](https://github.com/apache/pulsar/issues/9946)

> - Exposing prometheus metrics for Pulsar function local run mode [#10156](https://github.com/apache/pulsar/pull/10156)

Reviewer: ```suggestion
- Exposing Prometheus metrics for Pulsar function local run mode [#10156](https://github.com/apache/pulsar/pull/10156)
```

- Expose local run config metrics-port-start to CLI [#10185](https://github.com/apache/pulsar/pull/10185)
- Fix Pulsar Function localrun with multiple instances and metrics server is enabled [#10208](https://github.com/apache/pulsar/pull/10208)
- Improve localrun performance by using JVM options [#10273](https://github.com/apache/pulsar/pull/10273)

> - Fix function api can not use authdata to check superuser [#10364](https://github.com/apache/pulsar/pull/10364)

Reviewer: ```suggestion
- Fix function API can not use authdata to check superuser [#10364](https://github.com/apache/pulsar/pull/10364)
```

- Fix deadlock on Monitoring thread blocked by LeaderService.isLeader() [#10502](https://github.com/apache/pulsar/pull/10502)
- Allow user to set custom configs to plugin worker service [#10504](https://github.com/apache/pulsar/pull/10504)
- K8s Function Name Length Check Allows Invalid StatefulSet [#10531](https://github.com/apache/pulsar/pull/10531)
- Get function cluster from broker config when start function worker with broker [#10552](https://github.com/apache/pulsar/pull/10552)
- Process async results in the same Java runnable thread [#10618](https://github.com/apache/pulsar/pull/10618)
- Support using AutoProduceBytesSchema as the function output schema [#10716](https://github.com/apache/pulsar/pull/10716)
- ReflectionUtils use Class.forName in order to properly discover classes in Functions Runtime while using DefaultImplementation [#10827](https://github.com/apache/pulsar/pull/10827)
- Fix the out of index issue when dispatch messages based on the avgBatchSizePerMsg [#10828](https://github.com/apache/pulsar/pull/10828)

#### IO Connectors
- [Kafka Source Connector] Fix invalid topic name generation [#9035](https://github.com/apache/pulsar/pull/9035)

> - [Kafka Source Connector] Allow to manage Avro encoded messages [#9448](https://github.com/apache/pulsar/pull/9448)

Reviewer: ```suggestion
- [Kafka Source Connector] Allow managing Avro encoded messages [#9448](https://github.com/apache/pulsar/pull/9448)
```

> - Configure Pulsar IO connector yaml file generator for all connectors [#9629](https://github.com/apache/pulsar/pull/9629)

Reviewer: ```suggestion
- Configure Pulsar IO connector YAML file generator for all connectors [#9629](https://github.com/apache/pulsar/pull/9629)
```

- [Kafka Source Connector] KeyValue schema support for KafkaBytesSource [#10002](https://github.com/apache/pulsar/pull/10002)

> - Allow to use GenericObject for developing a Sink connector [#10034](https://github.com/apache/pulsar/pull/10034)

Reviewer: ```suggestion
- Allow using GenericObject for developing a Sink connector [#10034](https://github.com/apache/pulsar/pull/10034)
```

- Sink<GenericObject> unwrap internal AutoConsumeSchema and allow to handle topics with KeyValue schema [#10211](https://github.com/apache/pulsar/pull/10211)
- Using ObjectMapper instead of Gson to parse Source/SInk configuration [#10441](https://github.com/apache/pulsar/pull/10441)
- Expose subscription type in the SinkContext [#10446](https://github.com/apache/pulsar/pull/10446)

> - Add ability to seek/pause/resume for a topic to the SinkContext [#10498](https://github.com/apache/pulsar/pull/10498)

Reviewer: ```suggestion
- Add the ability to seek/pause/resume for a topic to the SinkContext [#10498](https://github.com/apache/pulsar/pull/10498)
```

- Only return persistent topic when list tables [#10368](https://github.com/apache/pulsar/pull/10368)
- Fix Presto startup on JDK11 [#10559](https://github.com/apache/pulsar/pull/10559)
- Fix pulsar sql issues when run select count(*) for the table with primary schema [#10840](https://github.com/apache/pulsar/pull/10840)

#### Tiered Storage
- Fix AWS credentials usages [#8950](https://github.com/apache/pulsar/pull/8950)
- Support ALI OSS tiered storage [#8985](https://github.com/apache/pulsar/pull/8985)

#### Proxy
- Fix the proxy does not support partitioned topic auto-creation type [#7903](https://github.com/apache/pulsar/issues/7903)
- Enable Conscrypt for Jetty in the Broker and in the Proxy [#10541](https://github.com/apache/pulsar/pull/10541)

#### Java Client
- Fix warn log on the producer side when duplicated messages have been dropped [#8729](https://github.com/apache/pulsar/pull/8729)
- Allow setting subscription name for Reader [#8801](https://github.com/apache/pulsar/pull/8801)
- Fix authParams showing in log with secret string(*****) [#8910](https://github.com/apache/pulsar/pull/8910)

> - Avoid enable DLQ on Key_Shared subscription [#9163](https://github.com/apache/pulsar/pull/9163)

Reviewer: enabling

- Support multi-topic reader [#9995](https://github.com/apache/pulsar/pull/9995)
- Process messages from different partitions on different listener threads [#10017](https://github.com/apache/pulsar/pull/10017)
- Ensure close resource to avoid memory leak [#10028](https://github.com/apache/pulsar/pull/10028)
- Support set start message ID for each topic/partition on Reader [#10033](https://github.com/apache/pulsar/pull/10033)
- Add a JSON RecordBuilder to the GenericJsonSchema [#10052](https://github.com/apache/pulsar/pull/10052)
- Allow GenericRecord to wrap any Java Object [#10057](https://github.com/apache/pulsar/pull/10057)
- Fix NPE while configuring consumer builder [#10063](https://github.com/apache/pulsar/pull/10063)
- Support get native schema [#10076](https://github.com/apache/pulsar/pull/10076)
- Support KeyValue on Message.getValue() when using AutoConsumeSchema [#10107](https://github.com/apache/pulsar/pull/10107)
- Change the default retry topic name and dead letter topic name [#10129](https://github.com/apache/pulsar/pull/10129)
- Fix KeyValue with SEPARATED encoding for the GenericObject [#10186](https://github.com/apache/pulsar/pull/10186)

> - Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)

Reviewer: ```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decode the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```
```suggestion
- Ensure download schema before decoding the payload for the AUTO_CONSUME schema [#10248](https://github.com/apache/pulsar/pull/10248)
```

- Fix batch message handling of the UnAckedMessageTracker [#9170](https://github.com/apache/pulsar/pull/9170)
- Fix ServerError is not converted to string in log [#9277](https://github.com/apache/pulsar/pull/9277)
- Remove Boost::System runtime dependency [#9498](https://github.com/apache/pulsar/pull/9498)
- Removed usages of boost::regex [#9533](https://github.com/apache/pulsar/pull/9533)
- Account for different variables names on different CMake versions [#9559](https://github.com/apache/pulsar/pull/9559)
- Allow to disable static or dynamic lib at build time [#9570](https://github.com/apache/pulsar/pull/9570)
- Avoid multiple compilations of same source files [#9675](https://github.com/apache/pulsar/pull/9675)
- Support configure debug level logs simply [#10031](https://github.com/apache/pulsar/pull/10031)
- Add /opt/homebrew/ as a possible path for OpenSSL on Mac [#10141](https://github.com/apache/pulsar/pull/10141)
- Fix race condition in MemoryLimitController [#10142](https://github.com/apache/pulsar/pull/10142)
- Fix releasing semaphore and memory quota after send timeout [#10144](https://github.com/apache/pulsar/pull/10144)

> - Allow to configure memory limit from C API [#10145](https://github.com/apache/pulsar/pull/10145)

Reviewer: ```suggestion
- Allow configuring memory limit from C API [#10145](https://github.com/apache/pulsar/pull/10145)
```
